### PR TITLE
hardshrink_cpu and hardshrink_backward_cpu refactoring with at::native::cpu_kernel

### PR DESCRIPTION
**Pull request by pbelevich**

### CODE REVIEW

**Comment by fmassa** [comment]
We can potentially get better performance tha is to vectorization if you move this code to the `cpu` folder, see binary ops there for an example

**Comment by VitalyFedyunin** [comment]
Please include performance comparison and measurements strategy as part of PR description.

File: aten/src/ATen/native/Activation.cpp [hunk 1]
```diff
@@ -5,6 +5,7 @@
 #include <ATen/Dispatch.h>
 #include <ATen/NativeFunctions.h>
 #include <ATen/native/TensorIterator.h>
+#include <ATen/native/cpu/Loops.h>
```
Comment (fmassa) [inline aten/src/ATen/native/Activation.cpp]: nit: is this header needed, now that you have moved the loops to the `cpu/` folder?

**Comment by pbelevich** [comment]
CPU_tensor_apply2: 4.88 ms ± 213 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
cpu_kernel:         223 µs ± 2.72 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)
cpu_kernel_vec:    40.1 µs ± 7.64 µs per loop (mean ± std. dev. of 7 runs, 10000 loops each)

cpu_kernel_vec gives 100x performance boost

File: aten/src/ATen/native/cpu/Activation.cpp [hunk 2]
```diff
@@ -175,11 +175,41 @@ void GeluBackwardKernelImpl(const Tensor& dY, const Tensor& X, Tensor* dX) {
...
+
+void hardshrink_backward_cpu_kernel(TensorIterator& iter, Scalar lambd) {
+  AT_DISPATCH_FLOATING_TYPES(iter.dtype(), "hardshrink_backward_cpu", [&] {
+    auto lambd_val = lambd.to<scalar_t>();
+    cpu_kernel_vec(iter,
+      [=](scalar_t grad_val, scalar_t self_val) {
+        return (self_val >= -lambd_val && self_val <= lambd_val) ? scalar_t(0) : grad_val;
+      },
+      [=](Vec256<scalar_t> grad_val, Vec256<scalar_t> self_val) {
+        return ((self_val < -lambd_val) | (self_val > lambd_val)) & grad_val;
```
Comment (VitalyFedyunin) [inline aten/src/ATen/native/cpu/Activation.cpp]: Im not sure we support & for such use case, can you please write a test that compares internal implementation and manual python iteration over numel > 50000k 1D tensor.

**Review by VitalyFedyunin — CHANGES_REQUESTED** [review]
Please undo submodule changes

File: aten/src/ATen/native/Activation.cpp [hunk 3]
```diff
@@ -339,35 +341,15 @@ std::tuple<Tensor, Tensor> prelu_backward_cpu(const Tensor& grad_out_, const Ten
...
-      self,
-      out_tensor,
-      [&](
-        scalar_t& self_val,
-        scalar_t& out_tensor_val) {
-          out_tensor_val = (self_val >= -lambd_val && self_val <= lambd_val) ? scalar_t(0) : self_val;
-    });
-  });
+  auto iter = TensorIterator::unary_op(out_tensor, self);
+  hardshrink_cpu_stub(iter->device_type(), *iter, lambd);
```
Comment (VitalyFedyunin) [inline aten/src/ATen/native/Activation.cpp]: `hardshrink_cpu_stub(kCPU, *iter, lambd);` just to be sure, as we are in the CPU function

File: aten/src/ATen/native/Activation.cpp [hunk 4]
```diff
@@ -339,35 +341,15 @@ std::tuple<Tensor, Tensor> prelu_backward_cpu(const Tensor& grad_out_, const Ten
...
-      out_tensor,
-      [&](
-        scalar_t& self_val,
-        scalar_t& grad_val,
-        scalar_t& out_tensor_val) {
-          out_tensor_val = (self_val >= -lambd_val && self_val <= lambd_val) ? scalar_t(0) : grad_val;
-    });
-  });
+  auto iter = TensorIterator::binary_op(out_tensor, grad, self);
+  hardshrink_backward_cpu_stub(iter->device_type(), *iter, lambd);
```
Comment (VitalyFedyunin) [inline aten/src/ATen/native/Activation.cpp]: `hardshrink_cpu_stub(kCPU, *iter, lambd);` just to be sure, as we are in the CPU function

**Review by VitalyFedyunin — APPROVED** [review]
Everything looks good, you may or may not apply additional changes request before landing.

**Pull request closed without merging**